bson/unmarshal_test.go

```diff
@@ -278,3 +278,158 @@ func TestUnmarshalExtJSONWithUndefinedField(t *testing.T) {
 		})
 	}
 }
+
+func TestUnmarshalBSONWithUndefinedField(t *testing.T) {
+	// When unmarshalling BSON, fields that are undefined in the destination struct are skipped.
+	// This process must not skip other, defined fields and must not raise errors.
+	type expectedResponse struct {
+		DefinedField string `bson:"DefinedField"`
+	}
+
+	unmarshalExpectedResponse := func(t *testing.T, BSON []byte) *expectedResponse {
```

Comment:
What you have now is perfectly fine, but I wanted to point out that you can reduce `unmarshalExpectedResponse` and `marshalBSON` to a single function:

```
createExpectedResponse := func(t *testing.T,  doc D) *expectedResponse
```

This would internally call `Marshal` followed by `Unmarshal`.

Reply:
Ah great point! Combined.